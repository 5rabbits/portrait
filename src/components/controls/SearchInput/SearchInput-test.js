import { React, toJSON, mount } from 'test/helper'
import SearchInput from 'controls/SearchInput'

describe('SearchInput', () => {
  it('should render with minimum props', () => {
    const component = mount(<SearchInput />)

    expect(toJSON(component)).toMatchSnapshot()
  })

  describe('setValue(value)', () => {
    describe('if props.onChange is defined', () => {
      it('should call props.onChange passing the new value', () => {
        const handleChange = jest.fn()
        const component = mount(<SearchInput onChange={handleChange} />)
        const instance = component.instance()

        instance.setValue('test')

        expect(handleChange).toHaveBeenCalledTimes(1)
        expect(handleChange).toHaveBeenCalledWith('test')
      })
    })

    describe('if props.onChange is undefined', () => {
      it('should not fail', () => {
        const component = mount(<SearchInput />)
        const instance = component.instance()

        expect(() => {
          instance.setValue('test')
        }).not.toThrow()
      })
    })

    describe('if props.value is != ""', () => {
      it('should not call to onChange', () => {
        const onClick = jest.fn()
        const component = mount(
          <SearchInput
            value="test"
            onChange={onClick}
          />,
        )
        const instance = component.instance()

        instance.setValue('aa')

        expect(onClick).not.toHaveBeenCalled()
      })
    })
  })

  describe('focus()', () => {
    it('should call focus on the underlying input', () => {
      const component = mount(<SearchInput />)
      const instance = component.instance()
      const spy = jest.fn()

      instance.input.focus = spy
      instance.focus()

      expect(spy).toHaveBeenCalledTimes(1)
    })
  })

  describe('handleClearClick(event)', () => {
    it('should prevent the event default action', () => {
      const event = { preventDefault: jest.fn() }
      const component = mount(<SearchInput />)
      const instance = component.instance()

      instance.handleClearClick(event)

      expect(event.preventDefault).toHaveBeenCalledTimes(1)
    })

    it('should call setValue with an empty string', () => {
      const event = { preventDefault: jest.fn() }
      const component = mount(<SearchInput />)
      const instance = component.instance()
      const spy = jest.fn()

      instance.setValue = spy
      instance.handleClearClick(event)

      expect(spy).toHaveBeenCalledTimes(1)
      expect(spy).toHaveBeenCalledWith('')
    })

    it('should call focus', () => {
      const event = { preventDefault: jest.fn() }
      const component = mount(<SearchInput />)
      const instance = component.instance()
      const spy = jest.fn()

      instance.focus = spy
      instance.handleClearClick(event)

      expect(spy).toHaveBeenCalledTimes(1)
    })

    describe('if props.onClear is defined', () => {
      it('should call props.onClear', () => {
        const event = { preventDefault: jest.fn() }
        const handleClear = jest.fn()
        const component = mount(<SearchInput onClear={handleClear} />)
        const instance = component.instance()

        instance.handleClearClick(event)

        expect(handleClear).toHaveBeenCalledTimes(1)
      })
    })

    describe('if props.onClear is undefined', () => {
      it('should call props.onClear', () => {
        const event = { preventDefault: jest.fn() }
        const component = mount(<SearchInput />)
        const instance = component.instance()

        expect(() => {
          instance.handleClearClick(event)
        }).not.toThrow()
      })
    })
  })

  describe('handleChange(event)', () => {
    it('should call setValue passing the event target value', () => {
      const event = { target: { value: 'test' } }
      const component = mount(<SearchInput />)
      const instance = component.instance()
      const spy = jest.fn()

      instance.setValue = spy
      instance.handleChange(event)

      expect(spy).toHaveBeenCalledTimes(1)
      expect(spy).toHaveBeenCalledWith('test')
    })
  })

  describe('handleKeyDown(event)', () => {
    describe('if props.onSubmit is defined', () => {
      describe('if the pressed key is ENTER', () => {
        it('should call props.onSubmit passing the current value', () => {
          const event = { key: 'Enter' }
          const handleSubmit = jest.fn()
          const component = mount(<SearchInput onSubmit={handleSubmit} />)
          const instance = component.instance()

          component.setState({ value: 'test' })
          instance.handleKeyDown(event)

          expect(handleSubmit).toHaveBeenCalledTimes(1)
          expect(handleSubmit).toHaveBeenCalledWith('test')
        })
      })

      describe('if the pressed key is not ENTER', () => {
        it('should not call props.onSubmit', () => {
          const event = { key: 'A' }
          const handleSubmit = jest.fn()
          const component = mount(<SearchInput onSubmit={handleSubmit} />)
          const instance = component.instance()

          instance.handleKeyDown(event)

          expect(handleSubmit).not.toHaveBeenCalled()
        })
      })
    })

    describe('if props.onSubmit is undefined', () => {
      it('should not fail', () => {
        const event = { key: 'Enter' }
        const component = mount(<SearchInput />)
        const instance = component.instance()

        expect(() => {
          instance.handleKeyDown(event)
        }).not.toThrow()
      })
    })
  })
})
