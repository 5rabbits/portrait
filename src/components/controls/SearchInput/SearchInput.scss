@import '~core/variables';

.SearchInput {
  position: relative;
}

.SearchInput__input.form-control {
  width: 100%;
  box-shadow: none;
  transition: border-color 200ms ease-out, box-shadow 200ms ease-out;
}

.SearchInput__icon {
  position: absolute;
}

.SearchInput__clear {
  font-size: 20px;
  position: absolute;
  display: block;
  height: 20px;
  width: 20px;
  text-align: center;
  line-height: 18px;
  opacity: 0;
  transition: opacity 200ms ease-out;
  pointer-events: none;
  cursor: pointer;

  &:hover {
    text-decoration: none;
    color: darken(#A7B3BE, 10%);
  }
}

.SearchInput--with-value {
  .SearchInput__clear {
    pointer-events: auto;
    opacity: 1;
  }
}

.SearchInput--default {
  .SearchInput__input.form-control {
    background-color: #fff;
    border-color: $default-input-border-color;
    height: 36px;
    border-radius: 18px;
    padding-right: 40px;
    padding-left: 40px;

    &::placeholder {
      color: $placeholder-color;
    }

    &:focus {
      border-color: rgba(102, 175, 233, .4);
      box-shadow: 0 0 8px rgba(102, 175, 233, .6);
    }
  }

  .SearchInput__icon {
    color: #BAC3CC;
    margin-top: 10px;
    margin-left: 16px;
  }

  .SearchInput__clear {
    color: #BAC3CC;
    top: 7px;
    right: 12px;
    font-size: 26px;

    &:hover {
      color: darken(#BAC3CC, 10%);
    }
  }
}

.SearchInput--alternate {
  .SearchInput__input.form-control {
    background-color: #EEF1F3;
    border-color: #EEF1F3;
    height: 28px;
    border-radius: 14px;
    padding-right: 32px;
    padding-left: 32px;

    &::placeholder {
      color: #b2bdc7;
    }

    &:focus {
      border-color: rgba(102, 175, 233, .4);
      box-shadow: 0 0 8px rgba(102, 175, 233, .6);
    }
  }

  .SearchInput__icon {
    color: #b2bdc7;
    margin-top: 6px;
    margin-left: 12px;
  }

  .SearchInput__clear {
    color: #b2bdc7;
    top: 4px;
    right: 8px;

    &:hover {
      color: darken(#A7B3BE, 10%);
    }
  }
}
