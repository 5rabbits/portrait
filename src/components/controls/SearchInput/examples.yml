-
  description: Using `onChange` for real time search.
  code: >
    <SearchInput
      onChange={value => console.log(value)}
      placeholder="Write to search"
      style={{ width: '400px' }}
    />

-
  description: Using `onSubmit` and `onClear` to search when the user presses ENTER.
  code: >
    <SearchInput
      onClear={value => console.log('cleared')}
      onSubmit={value => console.log('submited:', value)}
      placeholder="Write to search"
      style={{ width: '400px' }}
    />

-
  description: Alternate theme.
  code: >
    <div style={{ padding: 16, background: '#fff' }}>
      <SearchInput
        placeholder="Write to search"
        style={{ width: '400px' }}
        theme="alternate"
      />
    </div>

-
  description: Alternate theme.
  code: >
    <div style={{ padding: 16, background: '#fff' }}>
      <SearchInput
        placeholder="Write to search"
        style={{ width: '400px' }}
        theme="alternate"
        value='test'
      />
    </div>
