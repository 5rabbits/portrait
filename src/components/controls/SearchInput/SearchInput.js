import React, { Component } from 'react'
import PropTypes from 'prop-types'
import cx from 'classnames'
import './SearchInput.scss'

/**
 * Input field that displays an icon and a clear button, intended to be used as a global search on
 * sections and popovers.
 */
export default class SearchInput extends Component {
  static propTypes = {
    /**
     * CSS class assigned to the input wrapper.
     */
    className: PropTypes.string,

    /**
     * Invoked when the input text changes, either by writing or by clicking the clear button.
     * The first argument is the current text.
     */
    onChange: PropTypes.func,

    /**
     * Invoked when the user clicks on the clear button.
     */
    onClear: PropTypes.func,

    /**
     * Invoked when the user presses the ENTER key. The first argument is the current text.
     */
    onSubmit: PropTypes.func,

    /**
     * Graphical theme of the input. `default` is intended for global section search and
     * `alternate` for popovers.
     */
    theme: PropTypes.oneOf(['default', 'alternate']).isRequired,
  }

  static defaultProps = {
    className: null,
    onChange: null,
    onClear: null,
    onSubmit: null,
    theme: 'default',
  }

  state = {
    value: '',
  }

  setValue = value => {
    this.setState({ value })

    if (this.props.onChange) {
      this.props.onChange(value)
    }
  }

  focus = () => {
    this.input.focus()
  }

  inputRef = input => {
    this.input = input
  }

  handleClearClick = event => {
    event.preventDefault()
    this.setValue('')
    this.focus()

    if (this.props.onClear) {
      this.props.onClear()
    }
  }

  handleChange = event => {
    this.setValue(event.target.value)
  }

  handleKeyDown = event => {
    if (this.props.onSubmit && event.key === 'Enter') {
      this.props.onSubmit(this.state.value)
    }
  }

  render() {
    const { className, theme, ...other } = this.props
    const { value } = this.state

    delete other.onClear
    delete other.onSubmit

    return (
      <div
        className={cx('SearchInput', className, `SearchInput--${theme}`, {
          'SearchInput--with-value': value,
        })}
        >
        <i className="fa fa-search SearchInput__icon" />

        <input
          {...other}
          className={cx('form-control', 'SearchInput__input')}
          onChange={this.handleChange}
          onKeyDown={this.handleKeyDown}
          ref={this.inputRef}
          type="text"
          value={value}
        />

        <a
          className="SearchInput__clear"
          href="#"
          onClick={this.handleClearClick}
          >
          &times;
        </a>
      </div>
    )
  }
}
