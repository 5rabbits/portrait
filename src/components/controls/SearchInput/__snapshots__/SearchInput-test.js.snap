// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput should render with minimum props 1`] = `
<SearchInput
  className={null}
  defaultValue=""
  onChange={null}
  onClear={null}
  onSubmit={null}
  theme="default"
>
  <div
    className="SearchInput SearchInput--default"
  >
    <i
      className="fa fa-search SearchInput__icon"
    />
    <input
      className="form-control SearchInput__input"
      defaultValue=""
      onChange={[Function]}
      onKeyDown={[Function]}
      type="text"
      value=""
    />
    <a
      className="SearchInput__clear"
      href="#"
      onClick={[Function]}
    >
      Ã—
    </a>
  </div>
</SearchInput>
`;
